spec.md

```diff
@@ -0,0 +1,955 @@
+
+The canonical file extension for textfiles is `.txt`.
+
+Songs consist of a header and a body.
+The header contains metadata about the song.
+The body contains the musical data.
+A file SHOULD end with an `E` on a single line.
+Everything after a trailing `E` MUST be ignored.
+
+> [!CAUTION]
+>
+> Whether the trailing `E` is optional or not is open for discussion ([#44](https://github.com/UltraStar-Deluxe/format/issues/44)).
+
+```abnf
+song = file-header
+       file-body
+       [ %x45 *char ]  ; E
+char = %x00-10FFFF
+```
+
+### 2.1. Line Endings and Whitespace
+
+Both the header and the body of a file are defined in terms of lines.
+A line is a string of text that is terminated with an end-of-line sequence.
+
+```abnf
+end-of-line = ( CR / LF / CRLF )
+empty-line  = end-of-line
+```
+
+Implementations SHOULD use a single Line Feed (`%x0A`) as line terminator.
+Implementations MUST accept the end of input (`EOF`) as a valid line terminator.
+Empty lines are ignored throughout the entire file.
+
+> [!CAUTION]
+>
+> The allowed end-of-line sequences as well as potential recommendations for implementations are currently open for discussion ([#43](https://github.com/UltraStar-Deluxe/format/issues/43)).
+
+> [!CAUTION]
+>
+> Whether empty lines are allowed or not is currently open for discussion ([#43](https://github.com/UltraStar-Deluxe/format/issues/43)).
+> Whether a line that consists only of whitespace is recognized as an empty line has not been decided yet.
+
+Whitespace is used as a separator in many places of the format.
+Only space (`%x20`) and horizontal tab (`%x09`) are recognized as whitespace characters.
+In particular other unicode characters with the property `White_Space=yes` MUST NOT be treated as whitespace characters in the context of this specification.
+
+```abnf
+WSP = ( SP / HTAB )
+```
+
+> [!CAUTION]
+>
+> Definition and handling of whitespace characters is currently open for discussion ([#46](https://github.com/UltraStar-Deluxe/format/issues/46)).
+
+## 2. The File Header
+
+The header of a song consists of a sequence of key-value pairs.
+Each line in the header section starts with a hash.
+The key and value of a header are separated by a colon.
+Whitespace around key and value is ignored.
+
+> [!CAUTION]
+>
+> The terminology “header” is not decided yet.
+> Other terms used for the same concept are “tag”, “attribute” or “field”.
+
+```abnf
+file-header  = *( header / empty-line )
+header       = hash *WSP header-key *WSP colon *WSP header-value *WSP line-break
+header-key   = header-char
+               [ *( WSP / header-char )
+                 header-char ]
+header-value = single-value / multi-value
+multi-value  = single-value [ comma single-value ] [ comma ]
+single-value = *( header-char / colon )
+
+header-char  = %x00-09 /  ; exclude line feed
+               %x0B-0C /  ; exclude carriage return
+               %x0E-39 /  ; exclude colon
+               %x3B-10FFFF
```

Comment:
I think at least USDX allows spaces in header keys and values (although there are no standard headers that include whitespace). I think it makes sense to restrict the character set allowed for header keys (either non-whitespace or maybe even just ASCII) but I feel like that should be discussed in a separate issue.

I have added a caution marker that this is not decided yet.

Reply:
Sorry, I wasn't clear. I was referring to `header-key`: Since `WSP` is a subset of `header-char`, `header-char [ *( WSP / header-char ) header-char ]` is just the same as `1*header-char`. That's why I assumed you meant to exclude `WSP` from `header-char`. Then the definition of `header-key` would make more sense as it would ensure the first and last characters are not whitespace.

Reply:
Ah yeah I see. Originally I tried to exclude whitespace from the start and end of header keys but I felt the grammar got unnecessarily complicated so I put the part about leading and trailing whitespace in the text part. I must have forgotten to update the `header-key` when I made that change. I have now replaced it with `header-key = 1*header-char`.

Do you have a preference whether we should try handle these whitespace scenarios in the grammar? I think I prefer the current simplicity but I'd be open to change it for the sake of clarity.